File: dofbot_python/ctrl_all_servo.py
"""Sweep all servos through their full range.

Moves every joint back and forth so you can observe coordinated motion and
verify wiring/power without editing the script manually.
"""

from Arm_Lib import Arm_Device
import time

from cli_args import parse_common_args


def ctrl_all_servo(arm: Arm_Device, angle: int, s_time: int = 500) -> None:
    """Command every servo to the supplied angle with a shared duration."""
    arm.Arm_serial_servo_write6(
        angle, 180 - angle, angle, angle, angle, angle, s_time
    )
    # Give the hardware time to complete the motion before the next step.
    time.sleep(s_time / 1000)


def run_sequence(arm: Arm_Device) -> None:
    """Continuously sweep the servos until interrupted."""
    dir_state = 1
    angle = 90
    arm.Arm_serial_servo_write6(90, 90, 90, 90, 90, 90, 500)
    time.sleep(1)

    while True:
        if dir_state == 1:
            angle += 1
            if angle >= 180:
                dir_state = 0
        else:
            angle -= 1
            if angle <= 0:
                dir_state = 1

        # Move all servos in small increments to create a smooth sweep.
        ctrl_all_servo(arm, angle, 10)
        time.sleep(10 / 1000)


def main() -> None:
    args = parse_common_args(__doc__)
    arm = Arm_Device(args.port)
    try:
        time.sleep(args.init_delay)
        try:
            run_sequence(arm)
        except KeyboardInterrupt:
            print("Program closed!")
    finally:
        del arm


if __name__ == "__main__":
    main()

File: dofbot_c++/ctrl_all_servo.cpp
/**
 * @file ctrl_all_servo.cpp
 * @brief Sweeps all servos across their range, mirroring the Python demo.
 */

#include "Arm_Lib.h"
#include "cli_args.h"

#include <atomic>
#include <chrono>
#include <csignal>
#include <exception>
#include <iostream>
#include <thread>

namespace {
    std::atomic<bool> g_running(true);

    void handle_signal(int signum) {
        if (signum == SIGINT) {
            g_running = false;
        }
    }

    void ctrl_all_servo(Arm_Device& arm, int angle, int move_time_ms) {
        arm.Arm_serial_servo_write6(angle, 180 - angle, angle, angle, angle, angle, move_time_ms);
        if (!g_running) {
            return;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(move_time_ms));
    }
}

int main(int argc, char* argv[]) {
    const std::string description =
        "Continuously sweep every joint back and forth to verify coordinated motion.";

    std::signal(SIGINT, handle_signal);

    try {
        const CommonArgs args = parse_common_args(argc, argv, description);
        Arm_Device arm(args.port);

        std::this_thread::sleep_for(std::chrono::duration<double>(args.init_delay));

        int angle = 90;
        int dir_state = 1;

        std::cout << "Initializing pose..." << std::endl;
        arm.Arm_serial_servo_write6(90, 90, 90, 90, 90, 90, 500);
        std::this_thread::sleep_for(std::chrono::seconds(1));

        std::cout << "Sweeping servos. Press Ctrl+C to stop." << std::endl;
        while (g_running) {
            if (dir_state == 1) {
                ++angle;
                if (angle >= 180) {
                    dir_state = 0;
                }
            } else {
                --angle;
                if (angle <= 0) {
                    dir_state = 1;
                }
            }

            ctrl_all_servo(arm, angle, 10);
            if (!g_running) {
                break;
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(10));
        }

        std::cout << "\nProgram closed." << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}
